lib/src/relations/to_many.dart

```diff
@@ -0,0 +1,271 @@
+import '../store.dart';
+
+/// Manages a to-many relation, an unidirectional link from a "source" entity to
+/// multiple objects of a "target" entity.
+///
+/// You can:
+///   - [add()] new objects to the relation.
+///   - [removeAt()] target objects from the relation at a specific list index.
+///   - [remove(id:)] target objects from the relation by an ID.
+///
+/// ```
+/// class Student {
+///   final teachers = ToMany<Teacher>();
+///   ...
+/// }
+///
+/// // Example 1: create a relation
+/// final teacher1 = Teacher();
+/// final teacher2 = Teacher();
+///
+/// final student1 = Student();
+/// student1.teachers.add(teacher1);
+/// student1.teachers.add(teacher2);
+///
+/// final student2 = Customer();
+/// student2.teachers.add(teacher2);
+///
+/// // saves students as well as teachers in the database
+/// store.box<Student>().putMany([student1, student2]);
+///
+///
+/// // Example 2: remove a relation
+///
+/// student.teachers.removeAt(index)
+/// student.teachers.applyToDb(); // or store.box<Student>().put(student);
+/// ```
+class ToMany<EntityT> extends Object with ListMixin<EntityT> {
+  /*late final*/ Store _store;
+
+  /*late final*/
+  Box<EntityT> _box;
+
+  /*late final*/
+  Box _srcBox;
+
+  /*late final*/
+  EntityDefinition<EntityT> _entity;
+
+  RelInfo _rel;
+
+  List<EntityT> /*?*/ __items;
+  final _counts = <EntityT, int>{};
+
+  @override
+  int get length => _items.length;
+
+  @override
+  set length(int newLength) {
+    if (newLength < _items.length) {
+      _items
+          .getRange(newLength, _items.length)
+          .forEach((EntityT element) => _track(element, -1));
+    }
+    _items.length = newLength;
+  }
+
+  @override
+  EntityT operator [](int index) => _items[index];
+
+  @override
+  void operator []=(int index, EntityT element) {
+    final items = _items;
+    if (0 > index || index >= items.length) {
+      throw RangeError.index(index, this);
+    }
+    if (index < items.length) {
+      if (items[index] == element) return;
+      _track(items[index], -1);
+    }
+    items[index] = element;
+    _track(element, 1);
+  }
+
+  @override
+  void add(EntityT element) {
+    if (element == null) ArgumentError.notNull('element');
+    // TODO add and addAll() don't need to load from DB, unless already loaded
+    _items.add(element);
+    _track(element, 1);
+  }
+
+  @override
+  void addAll(Iterable<EntityT> iterable) {
+    iterable.forEach((element) {
+      if (element == null) ArgumentError.notNull('element');
+      _track(element, 1);
+    });
+    _items.addAll(iterable);
+  }
+
+  // note: to override, arg must be "Object", same as in the base class.
+  @override
+  bool remove(Object /*?*/ element) {
+    if (!_items.remove(element)) return false;
+    if (element != null) _track(element, -1);
+    return true;
+  }
+
+  /// "add":    increment = 1
+  /// "remove": increment = -1
+  void _track(EntityT object, int increment) {
+    if (_counts.containsKey(object)) {
```

Comment:
Do you know how Dart computes object equality? Asking for a friend new with Dart.

As the tests work I guess this is fine. Just asking because in Java it's possible to override the equals calculation and was wondering if Dart can do the same.

Reply:
Each object instance has a unique integer hash code. you can override that if you want.

https://dart.dev/guides/libraries/library-tour#implementing-map-keys

Reply:
So this is just like Java, cool, thanks (did not find that)!